### ISSUE TITLE
Calling model.test_on_batch after model.evaluate returns corrupted values for the loss and the metrics

### ISSUE THREAD

**Issue by ogrisel** [tensorflow/tensorflow#47340]
**System information**
- Google colab with tf 2.4.1 (v2.4.1-0-g85c8b2a817f )
- with CPU or GPU runtimes, it does not matter

**Describe the current behavior**

Calling `model.test_on_batch` after calling `model.evaluate` gives incorrect results.

**Describe the expected behavior**

Calling `model.test_on_batch` should  return the a value that does not depend on whether`model.evaluate`  was called before.

**Standalone code to reproduce the issue**

Let's define a randomly initialized model evaluated on randomly generated data. If we call `model.test_on_batch` directly, everything is fine:

```python
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

rng = np.random.RandomState(0)
batch_size = 32
n_samples, n_features = batch_size * 10, 5
X = rng.normal(size=(n_samples, n_features))
y = rng.randint(low=0, high=2, size=X.shape[0])

model = Sequential([Dense(1, input_shape=(n_features,),
                          activation="sigmoid")])
model.compile(optimizer="adam", loss='binary_crossentropy',
              metrics=['accuracy'])

print("model.test_on_batch without model.evaluate")
for i in range(3):
    loss, acc = model.test_on_batch(X[:batch_size], y[:batch_size])
    print(loss, acc)
```

output:

```
model.test_on_batch without model.evaluate
6.294709205627441 0.5625
6.294709205627441 0.5625
6.294709205627441 0.5625
```

If we then call `evaluate`, the fist call of `model.test_on_batch` return an incorrect value:

```python
normal_loss, normal_acc = model.evaluate(
    X, y, batch_size=batch_size, verbose=0)

print("model.test_on_batch *after* model.evaluate")
for i in range(3):
    loss, acc = model.test_on_batch(X[:batch_size], y[:batch_size])
    print(loss, acc)
```

output:

```
model.test_on_batch *after* model.evaluate
6.585799694061279 0.4801136255264282
6.294709205627441 0.5625
6.294709205627441 0.5625
```

Note: this is a minimal reproducer of the report found in the comments of a similar standalone keras issue: https://github.com/keras-team/keras/issues/14086

It is probably also the cause or at least related to corrupted loss/metric values reported when using `evaluate_generator`:
https://github.com/keras-team/keras/issues/13780

**Comment by ogrisel** [comment]
The problem also happens if the `batch_size` argument of evaluate equal or greater than `X.shape[0]`.

**Comment by Saduf2019** [comment]
@ogrisel 
I ran your code on tf nightly and there is not much of a difference can you please refer to[ this link ](https://colab.research.google.com/gist/Saduf2019/09d1df811c8d2fc688079f63fd18bcf8/untitled548.ipynb)and let us know.

**Comment by ogrisel** [comment]
The difference is random because I did not seed the model parameters. But the bug is still there in nightly. For instance I re-ran your notebook and got:

```
model.test_on_batch *after* model.evaluate
6.819779872894287 0.4375
5.586616516113281 0.46875
5.586616516113281 0.46875
```

**Comment by sushreebarsa** [comment]
Was able to reproduce this issue in TF v2.5,please check the gist [here](https://colab.research.google.com/gist/sushreebarsa/a7702376cf6b787efc866332bc967775/untitled144.ipynb?authuser=1)..Thanks !

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.6.0-dev20210606,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/95e84ecec976207debf6513d1784ff8b/untitled166.ipynb?authuser=1) ..Thanks!

**Comment by bhack** [comment]
@ogrisel Can you confirm that `evaluate` corrupts also `train_on_batch`?

**Comment by bhack** [comment]
It seems I cannot reproduce this at https://github.com/keras-team/keras/pull/15342

**Comment by ogrisel** [comment]
I can reproduce on TF 2.6.0 on colab:

https://colab.research.google.com/drive/1NgS57SegqurxwgAu0t3R9N3TPudWJpo9?usp=sharing

**Comment by bhack** [comment]
@ogrisel Ok I can reproduce now on master at https://github.com/keras-team/keras/pull/15342

**Comment by mohantym** [comment]
Hi @ogrisel ! This issue is getting resolved in TF 2.7 . Attaching [Gist](https://colab.sandbox.google.com/gist/mohantym/6e36afd8ec501642367c885cc916cd11/tensorflow-gh-47340.ipynb#scrollTo=DyvDY7N5ipBo) for reference. Thanks!
